bin/curriculum/export/add_unit_source.rb

```diff
@@ -0,0 +1,110 @@
+    $options[:output_dir] = output_dir
+  end
+
+  opts.on("-h", "--help", "Prints this help") do
+    puts opts
+    exit
+  end
+end.parse!
+
+raise 'Input directory is required' unless $options[:input_dir] && !$options[:input_dir].to_s.strip.empty?
+
+$options[:output_dir] ||= $options[:input_dir]
+
+$output_dir = File.join('/mnt/tmp-curriculum-export', 'sourced', $options[:output_dir])
+FileUtils.mkdir_p($output_dir)
+
+require_relative '../../../deployment'
+
+start_time = Time.now
+puts "Loading Rails environment..."
+require_relative '../../../dashboard/config/environment'
+# load secret before parallelizing to reduce debug spew
+CDO.channels_api_secret
+puts "Rails environment loaded in: #{(Time.now - start_time).to_i} seconds"
+
+def list_s3_files(bucket, prefix)
+  s3 = Aws::S3::Client.new
+  response = s3.list_objects_v2(bucket: bucket, prefix: prefix)
+  keys = response.contents.map(&:key)
+  raise "No files found in s3://#{bucket}/#{prefix}/" if keys.empty?
+  keys
+end
+
+$max_processes = 25
+
+def process_s3_file(bucket, key)
+  output_filename = File.join($output_dir, File.basename(key))
+
+  if File.exist?(output_filename)
+    puts "Skipping #{File.basename(key)} because output file already exists: #{output_filename}"
+    return
+  end
+
+  puts "Downloading s3://#{bucket}/#{key}"
+  start_time = Time.now
+  s3 = Aws::S3::Client.new
+  response = s3.get_object(bucket: bucket, key: key)
+  rows = response.body.string.each_line.to_a
+  puts "Downloaded #{rows.size} rows in #{(Time.now - start_time).round(2)} seconds."
+
+  puts "Processing #{File.basename(key)} in parallel with #{$max_processes} processes"
+  start_time = Time.now
+  rows = Parallel.map(rows, in_processes: $max_processes) do |row|
+    data = JSON.parse(row)
+    channel_id = data['channel_id']
+    source = get_project_source(channel_id)
+    data['source'] = source
+    data.to_json
+  rescue JSON::ParserError => exception
+    puts "Error parsing JSON: #{exception}"
+    row
+  end
+
+  File.open(output_filename, 'w') do |file|
+    rows.each {|row| file.puts(row)}
+  end
+
+  puts "Processed #{rows.size} rows in #{(Time.now - start_time).round(2)} seconds."
+end
+
+def get_project_source(channel_id)
+  return nil unless channel_id
+
+  source_data = SourceBucket.new.get(channel_id, "main.json")
+  return nil unless source_data && source_data[:body] && source_data[:body].respond_to?(:string)
+
+  main_json = source_data[:body].string
+  JSON.parse(main_json)['source']
+rescue NoMethodError => exception
+  puts "Error getting source for channel id: #{channel_id}: #{exception}"
+  nil
+end
+
+def main
+  start_time = Time.now
+  keys = list_s3_files('cdo-data-sharing-internal', "unloaded-unit-progress/#{$options[:input_dir]}/")
+  puts "Found #{keys.size} files in s3://cdo-data-sharing-internal/unloaded-unit-progress/#{$options[:input_dir]}/"
```

Comment:
I'm a little confused, in the README the only previous step is `export_unit_progress.rb`, but `export_unit_progress.rb` doesn't seem to access S3 at all, let alone write to it. However, this step appears to get its list of things to work on from S3.

Is there a missing step here? Totally ignore if I'm just confused, which I prolly am.

Reply:
good point, there are hidden things happening here. `export_unit_progress.rb` invokes redshift, and the query uses `unload` to send the output directly to S3:
-  https://github.com/code-dot-org/code-dot-org/blob/c37476bf6d8c10da8b0a1528b4f898e2d51b4daa/bin/curriculum/export/export_unit_progress.sql.erb#L4
- https://github.com/code-dot-org/code-dot-org/blob/c37476bf6d8c10da8b0a1528b4f898e2d51b4daa/bin/curriculum/export/export_unit_progress.sql.erb#L240-L244

docs for `unload`: https://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD.html